quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -0,0 +1,823 @@
+/// Complex machinery for simplifying [`Expression`]s.
+use crate::expression::{
+    imag, real, Expression, ExpressionFunction, FunctionCallExpression, InfixExpression,
+    InfixOperator, PrefixExpression, PrefixOperator,
+};
+use std::cmp::min_by_key;
+
+/// Simplify an [`Expression`].
+pub(super) fn run(expression: &Expression) -> Expression {
+    simplify(expression, LIMIT)
+}
+
+/// Keep stack sizes under control
+const LIMIT: u64 = 10;
+
+/// Recursively simplify an [`Expression`] by hand, breaking into cases to make things more
+/// manageable.
+fn simplify(e: &Expression, limit: u64) -> Expression {
+    if limit == 0 {
+        // bail
+        return e.clone();
+    }
+    match e {
+        Expression::Address(_)
+        | Expression::Number(_)
+        | Expression::PiConstant
+        | Expression::Variable(_) => e.clone(),
```

Comment:
`expression/mod.rs` simplifies `PiConstant` to `Expression::Number(PI)` and down below you simplify `-pi` to `Expression::Number(-PI)` but here you don't convert to a Number.  Is this different by design?

Reply:
Not really! I just thought this whole class of things was constant and should simplify to themselves. Changing it to π => 3.14159… passes all tests, so I'll throw that in, too.